### PR TITLE
Use single php code base for multiple instances

### PR DESCRIPTION
**Pull request by JiCiT**
zoph.ini section names must match path section of REQUEST_URI without leading slash (/).

Maintaining a single zoph install for multiple instances seems "cleaner" than having multiple duplicate copies.  What I've done may not be the best implementation, but it does show the idea.  This change lets you set aliases in web server config pointing to the same physical file location.  Zoph.ini simply needs different DB info per install.

### CODE REVIEW

**Comment by jeroenrnl** [comment]
Hi Jason,

Good to hear from you again :-) Sorry it has taken a while to get back to you. I have not had much time to spend on Zoph recently and I was working on a big change that I have finally finished now.

Anyway, I like your idea, but there are two issues with it:
1. The original idea of "instances" was, that one could run several Zoph versions at the same time (I must confess, this was in a large part fixing my own problems as I usually run at least 2 versions concurrently - development and production - but often a few test versions as well.
2. The part where "php location" is used to distinguish the different installations is also used by the cli utility to find the location of the Zoph installation, because it needs to load the classes etc. 

Additionally, I'm not entirely sure what would happen if there are nested directories in the path. Currently, there are none - I think - but this may change in the future.

Finally, what about using different vhosts for different instances?

So, in short: I like the idea, but I don't think this implementation will work.

I have in the past thought of making Zoph some kind of multi-tennant, by running several version *from the same database*, but this is *way* more complicated than your proposal.

Maybe we could simply read **all** instances and if there are more than one with a matching php_location, *then* distinguish by (a part of) the URL? As said, I'm not sure if that should be the dirname, as I could imagine someone using various vhosts ("http://work.zoph.org/",  "http://private.zoph.org") for different installs...

Any thoughts?

**Comment by JiCiT** [comment]
Nice to be hear from you too. :-)  This is just off the top of me head....

Here's what my zoph.ini looks like:
```[zoph]
db_host = localhost
db_name = zoph
db_user = <username>
db_pass = <password>
db_prefix = zoph_
php_location = /usr/local/www/zoph

[zoph_alaina]
db_host = localhost
db_name = zoph_alaina
db_user = <username>
db_pass = <password>
db_prefix = zoph_
php_location = /usr/local/www/zoph
```
I *think* issues 1, 2, and the vhosts question can all be addressed in the same way.  For running different versions, you must be accessing them using different vhosts, or ports, or something, correct?  Within settings.inc.php, we could look at host, port, and the remainder of the URI.  Within zoph.ini, instead of using an arbitrary section name, we could set it like ``[<host>:<port>:<uri_path>]`` so it could match up for use in settings.inc.php.  Or the heading could remain arbitrary and then just add additional attributes within the section for host, port, and path and search through them within settings.ini.php.  We could also allow wild cards within each part.  For example, *.mydomain.com to match all hosts/sub-domains for mydomain.com.  Or simply * for the port to match all ports.  For each section, whatever php_location is appropriate/needed may be used depending on whether your trying to get multiple DBs/tenants to using a single code base or use multiple code bases against a single DB/tenant.

For a rudimentary multi-tenancy system, I think a single DB instance should work equally as well different DBs, *if* you use a different db_prefix for each instance.  I'm assuming that's what you're considering vs. somehow mingling multiple tenants within a single "set" of tables?  I'm also assuming that zoph.ini would be maintained at a "super admin" level?  That is, **not** by individual tenants.  As you said, more or less, lots and lots more thinking and decision making would be needed for anything more than the most basic multi-tenant system.

**Comment by jeroenrnl** [comment]
First, I think the header of the section should be something freeform, as this is what is used on the CLI to indicate which instance of Zoph you're talking to, such as
````zoph -i test --import *.JPG````
It would not be very userfriendly to require something like
````zoph -i http://test.zoph.org/test --import *.JPG````
there.
So I think we could add an *optional* url parameter. 
```
[test]
db_user= ...
...
url=http://test.zoph.org/test
```
I can imagine that sometimes a site is reached under different urls, for example if someone is self-hosting and doesn't have a real domain name. Fortunately ini-files in PHP can have arrays, so it could be something like this:
```
[test]
db_user= ...
...
url[]=http://myserver/zoph
url[]=http://zoph.dyndns.org
```
So, Zoph will work like this:
- it first tries to match from the php_location [unchanged from current behaviour]
- it the checks for a "url" setting in the ini section
- if it is an array, it checks them one by one
- as soon as a match is found, it picks that one an continues
- if no match is found, it skips to the next section, again tries matching on php_location first and if it matches, see above.

Some additional thoughts:
* settings.inc.php mixes class and global code, this should be split up
* settings.inc.php is a class file, but not in the classes dir.
* ever since I've introduced the conf::* classes I've not been very keen on the name of the settings class as it's confusing. I think it may be used by the cli to store some of the cli-parameters as well.
* settings.inc.php is currently not covered by unittests, this could be an opportunity to do that.
* documentation https://en.wikibooks.org/wiki/Zoph/Configuration/zoph.ini

I can imagine that the first 4 points a a little overwhelming if you're not that familiar with the Zoph code, especially because it will need some architecture decisions. Feel free to leave those to me. If you do want to make changes to those, please discuss first, as I have some architecture "rules" in my head that I have not yet documented anywhere.

**Comment by JiCiT** [comment]
That all sounds good except the order in which the matches are made.  If the php_location is matched first, that doesn't address my use case.  I have two ini sections using the same php_location, so they'd both match (or I'd guess the first listed section in the ini would actually be the one that would always match).  Maybe the match order could be a configuration option in the global section of the ini?  And it's default would be to conform to the current behavior.

**Comment by jeroenrnl** [comment]
It would work in your situation, because it would match on php_location first, but since it won't find a matching URL, it will then skip to the next php_location.

**Comment by JiCiT** [comment]
Ahh... OK.  That works then.  Thanks.

**Comment by jeroenrnl** [comment]
Hi @JiCiT I'm not sure if you're still working on this, but I've moved Zoph to [Gitlab](https://gitlab.com/jeroenrnl/zoph), please continue [here](https://gitlab.com/jeroenrnl/zoph/pull/96)

**Comment by jeroenrnl** [comment]
@JiCiT I haven't heard from you and you don't seem to have a Gitlab account, could you let me know if you still want to work on this?
